lib/prioritized-attributes.js

```diff
@@ -0,0 +1,171 @@
+  }
+
+  get(dest) {
+    const attrs = Object.create(null)
+
+    for (let [key, attr] of this.attributes) {
+      if (!(attr.destinations & dest)) {
+        continue
+      }
+
+      attrs[key] = typeof attr.value === 'string' && !attr.truncateExempt
+        ? byteUtils.truncate(attr.value, 255)
+        : attr.value
+    }
+
+    return attrs
+  }
+
+  has(key) {
+    this.attributes.has(key)
+  }
+
+  reset() {
+    this.attributes = new Map()
+  }
+
+  addAttribute(
+    destinations,
+    key,
+    value,
+    truncateExempt = false,
+    priority = ATTRIBUTE_PRIORITY.HIGH
+  ) {
+    const existingAttribute = this.attributes.get(key)
+
+    let hasDroppableAttributeKey = null
+    if (!existingAttribute && this.attributes.size === this.limit) {
+      hasDroppableAttributeKey = this._hasDroppableAttribute(priority)
+
+      if (!hasDroppableAttributeKey) {
+        logger.debug(
+          `Maximum number of custom attributes have been added.
+          Dropping attribute ${key} with ${value} type.`
+        )
+
+        return
+      }
+    }
+
+    if (existingAttribute && priority < existingAttribute.priority) {
+      logger.debug('incoming priority for \'%s\' is lower than existing, not updating.', key)
+      logger.trace(
+        '%s attribute retained value: %s, ignored value: %s',
+        key,
+        existingAttribute.value,
+        value
+      )
+      return
+    }
+
+    if (!isValidType(value)) {
+      logger.debug(
+        'Not adding attribute %s with %s value type. This is expected for undefined' +
+        'attributes and only an issue if an attribute is not expected to be undefined' +
+        'or not of the type expected.',
+        key,
+        typeof value
+      )
+      return
+    }
+
+    if (!this.isValidLength(key)) {
+      logger.warn(
+        'Length limit exceeded for attribute name, not adding: %s',
+        key
+      )
+      return
+    }
+
+
+    // Only set the attribute if at least one destination passed
+    const validDestinations = this.filter(destinations, key)
+    if (!validDestinations) {
+      return
+    }
+
+    if (hasDroppableAttributeKey) {
+      const droppableAttributeKey = this._lowPriorityAttributeNames.pop()
+      logger.trace(
+        'dropping existing lower priority attribute %s ' +
+        'to add higher priority attribute %s',
+        droppableAttributeKey,
+        key
+      )
+
+      this.attributes.delete(droppableAttributeKey)
+    }
+
+    this._set(validDestinations, key, value, truncateExempt, priority)
+  }
+
+  addAttributes(destinations, attrs) {
+    for (let key in attrs) {
+      if (properties.hasOwn(attrs, key)) {
+        this.addAttribute(destinations, key, attrs[key])
+      }
+    }
+  }
+
+  _hasDroppableAttribute(incomingPriority) {
+    // There will never be anything lower priority to drop
+    if (incomingPriority === ATTRIBUTE_PRIORITY.LOW) {
+      return
```

Comment:
I'm guessing this is the same as return false, low priority items will never having something that should be dropped in their favor.

Reply:
yeah. originally it wasn't a true/false function. i ideally should return false. maybe i'll do that and comment the other function a little more.